exp/query/api.go

```diff
@@ -88,44 +110,183 @@ func (a *API) Search(ctx context.Context, req *SearchRequest) (*SearchResponse,
+	dis := distancer.NewCosineDistanceProvider()
+	index, err := flat.New(flat.Config{
+		ID:               defaultLSMVectorsCompressedBucket,
+		DistanceProvider: dis,
+		RootPath:         localPath,
+	}, flatent.UserConfig{
+		BQ: bq,
+	}, store)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to initialize index: %w", err)
+	}
+	defer index.Shutdown(ctx)
+
+	// TODO(kavi): Here `limit` is not what you expect. It's the maxLimit.
+	// Currently `SearchByVectorDistance` api takes limit via `newSearchByDistParams(maxLimit)` which caller
+	// don't have control too.
+	certainty := 1 - threshold
+	matched_ids, distance, err := index.SearchByVectorDistance(vectors, certainty, int64(limit), nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to search by vector: %w", err)
+	}
+
+	opts := []lsmkv.BucketOption{
+		// lsmkv.WithStrategy(lsmkv.StrategyReplace),
+		lsmkv.WithSecondaryIndices(2),
+	}
+
+	if err := store.CreateOrLoadBucket(ctx, defaultLSMObjectsBucket, opts...); err != nil {
+		return nil, nil, fmt.Errorf("failed to vectorize query text: %w", err)
+	}
+
+	bkt := store.Bucket(defaultLSMObjectsBucket)
+
+	objs := make([]*storobj.Object, 0)
+
+	for _, id := range matched_ids {
+		key, err := indexDocIDToLSMKey(id)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to serialize ids returned from flat index: %w", err)
+		}
+		objB, err := bkt.GetBySecondary(0, key)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get object from store: %w", err)
+		}
+
+		obj, err := storobj.FromBinary(objB)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to decode object from store: %w", err)
+		}
+
+		objs = append(objs, obj)
+	}
+
+	return objs, distance, nil
+}
+
+func (a *API) loadOrDownloadLSM(ctx context.Context, collection, tenant string) (*lsmkv.Store, string, error) {
+	localPath := path.Join(a.offload.DataPath, strings.ToLower(collection), strings.ToLower(tenant), defaultLSMRoot)
+
+	// NOTE: Download only if path doesn't exist.
+	// Assumes, whatever in the path is latest.
+	// We will add a another way to keep this path upto date via some data versioning.
+	a.omu.Lock()
+	defer a.omu.Unlock()
+
+	_, err := os.Stat(localPath)
+	if os.IsNotExist(err) {
+		// src - s3://<collection>/<tenant>/<node>/
+		// dst (local) - <data-path/<collection>/<tenant>
+		if err := a.offload.Download(ctx, collection, tenant, nodeName); err != nil {
+			return nil, "", err
+		}
+	}
+
+	// TODO(kavi): Avoid creating store every time?
+	store, err := lsmkv.New(localPath, localPath, a.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
+	}
+	return store, localPath, nil
+}
+
 type SearchRequest struct {
 	Collection string
 	Tenant     string
+	NearText   []string // vector search
+	Certainty  float64  // threshold to match with certainty of vectors match
```

Comment:
Certainty is only a valid concept when the vector index is configured to use cosine distance. Does it make sense to support `distance` as well, for those collections whose distance metric is set to other than cosine? Or is this to be done at a later time?

Reply:
Ah now it explains why we have `distance` field in on SearchRequest protobuf.

```
type NearTextSearch struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	// protolint:disable:next REPEATED_FIELD_NAMES_PLURALIZED
	Query     []string             `protobuf:"bytes,1,rep,name=query,proto3" json:"query,omitempty"`
	Certainty *float64             `protobuf:"fixed64,2,opt,name=certainty,proto3,oneof" json:"certainty,omitempty"`
	Distance  *float64             `protobuf:"fixed64,3,opt,name=distance,proto3,oneof" json:"distance,omitempty"`
	MoveTo    *NearTextSearch_Move `protobuf:"bytes,4,opt,name=move_to,json=moveTo,proto3,oneof" json:"move_to,omitempty"`
	MoveAway  *NearTextSearch_Move `protobuf:"bytes,5,opt,name=move_away,json=moveAway,proto3,oneof" json:"move_away,omitempty"`
	// Deprecated: Marked as deprecated in v1/search_get.proto.
	TargetVectors []string `protobuf:"bytes,6,rep,name=target_vectors,json=targetVectors,proto3" json:"target_vectors,omitempty"` // deprecated in 1.26 - use targets
	Targets       *Targets `protobuf:"bytes,7,opt,name=targets,proto3" json:"targets,omitempty"`
}

```

I can add it in follow up PR if that's ok.